File: leetcode_75/Backtracking/python/216.py
from typing import List


# class Solution:
#     def combinationSum3(self, k: int, n: int) -> List[List[int]]:

#         def backtracking(cur_data, cur_num, target, result):
#             target -= cur_num
#             if target == 0 and len(cur_data) == k:
#                 cur_data = sorted(cur_data)
#                 if cur_data not in result:
#                     result.append(cur_data)

#             if target > 0 and len(cur_data) < k:
#                 for i in range(1, 10):
#                     if i not in cur_data:
#                         backtracking(cur_data + [i], i, target, result)

#             return result

#         return backtracking([], 0, n, [])


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def backtracking(start, path, target):
            if len(path) == k and target == 0:
                result.append(path[:])
                return

            if len(path) >= k or target < 0:
                return

            for i in range(start, 10):
                path.append(i)
                backtracking(i + 1, path, target - i)
                path.pop()

        result = []
        backtracking(1, [], n)
        return result

File: leetcode_75/Backtracking/cpp/216.cpp
#include <vector>

using namespace std;

class Solution
{

    vector<vector<int>> result;
    int k;

    void backtracking(int start, vector<int> path, int target)
    {
        if (path.size() == this->k && target == 0)
        {
            this->result.push_back(path);
            return;
        }

        if (path.size() >= this->k || target < 0)
        {
            return;
        }

        for (int i = start; i < 10; ++i)
        {
            path.push_back(i);
            backtracking(i + 1, path, target - i);
            path.pop_back();
        }
    }

public:
    vector<vector<int>> combinationSum3(int k, int n)
    {
        this->k = k;
        this->backtracking(1, {}, n);

        return this->result;
    }
};